### PR TITLE
handle will-download

### LINKED ISSUE #151: Cannot download a file
Has anyone been able to download a file?

I have tried a few approaches. The first was:

```js
.on('filePicker', function(oldFile) {
            console.log('onFilePicker(' + oldFile + ') called');
            return 'a-filename.ext';
          })
```

That code was never triggered.

Then [someone mentioned](https://groups.google.com/forum/#!topic/phantomjs/JChUakj--24) onFileDownload, but that also didn't work.

I have a listener to output resource requests (and it is correctly logging the attempt to retrieve the file):

```js
.on('resourceRequested', function(requestData, networkRequest) {
            console.log('Request (#' + requestData.id + '): ' + JSON.stringify(requestData));
          })
```

I can't just stream the file to disk because it's behind an authenticated session, hence the need for PhantomJS.

Anyone able to get file downloading to work? I think [Casper](http://docs.casperjs.org/en/latest/modules/casper.html#casper-download) has it going, so I will try that next.

### LINKED ISSUE #332: Prevent automatic downloads
When visiting a page that prompts a download window after the onload, is there a way to prevent it to pop up, so that I can use nightmarejs without needing to interact and close those windows?

Something like a preventDefault on a will-download event?

### PR DESCRIPTION
**Pull request by rosshinkley**
Allow users to automatically accept, automatically ignore, or handle downloads on a per-case basis. Fixes #332 and #371 with the addition of the `ignoreDownloads` option, should fix #151 with the `download` event and the various ways that can be used.

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -199,6 +210,30 @@ This event is triggered if `console.log` is used on the page. But this event is
...
+This event is triggered when Electron emits `'will-download'`.  This event is also emitted after downloads are started when [`DownloadItem`](https://github.com/atom/electron/blob/master/docs/api/download-item.md) emits `'updated'` or `'done'`.   The possible values for `event` are `'started'`, `'cancelled'`, `'interrupted'`, or `'completed'`.  Note that by listening to `'download'`, Nightmare expects the default download behavior to be overridden. 
+
+Downloads should be completed using `Nightmare.emit('download', action, downloadItem)`.  The possible values for `action` are `'cancel'`, `'continue'` for default behavior, or a file path (file name and extension inclusive) to save the download to an alternative location. The `downloadItem` parameter should use the item passed by `'download'`.
+
+For example:
+
+```javascript
+var nightmare = Nightmare();
+nightmare.on('download', function(event, downloadItem){
+  if(state == 'start'){
```
Comment (wcandillon) [inline Readme.md]: the state variable doesn't seem to be defined in this example.

File: Readme.md [hunk 2]
```diff
@@ -116,6 +116,12 @@ var nightmare = Nightmare({
 });
 ```
 
+##### ignoreDownloads
+Defines whether or not downloads should be ignored.
+
+##### downloadTimeout
+This will throw an eception if the `.wait('downloads-complete')` didn't return `true` within the set timeframe.
```
Comment (wcandillon) [inline Readme.md]: exception*

**Comment by wcandillon** [comment]
I reviewed it and tested it. It works great. Thank you so much for doing this.

File: Readme.md [hunk 3]
```diff
@@ -178,6 +184,11 @@ Wait until the element `selector` is present e.g. `.wait('#pay-button')`
 #### .wait(fn[, arg1, arg2,...])
 Wait until the `fn` evaluated on the page with `arg1, arg2,...` returns `true`. All the `args` are optional. See `.evaluate()` for usage.
 
+#### .wait('downloads-complete')
+Wait until all downloads are in a state of `'completed'`, `'interrupted'`, or `'cancelled'`
```
Comment (matthewmueller) [inline Readme.md]: i don't think it's a good idea to overload `nightmare.wait(str)` in this way. It sounds great semantically, but if we have custom tags like `<download-complete></download-complete>`, we'll run into problems.

Comment (rosshinkley) [inline Readme.md]: Fair enough.  What do you propose as an alternative?  Should there be a standalone method, eg `.waitForDownloads()`, or should the wait string start with a character that is illegal for tags, eg `.wait(':downloads-complete')`?

File: Readme.md (hunk 2 above)
Comment (rosshinkley) [inline Readme.md]: Good eye, thank you. :)

File: Readme.md (hunk 1 above)
Comment (rosshinkley) [inline Readme.md]: Ah, man, I missed the rename.  It should be `state == 'started'`, with `event` renamed to `state`.  The event name is `download`, and I wanted to be clear about what the first parameter was.

Also, `'start'` should be `'started'`.  Fix coming.  Thanks for the find!

**Comment by matthewmueller** [comment]
thanks for putting this together and i appreciate all the work that was put into this. currently this approach adds too much complexity to the API for solving this issue. 

i've thought about this a little bit and i think a better approach might be:

1. By default ignore all downloads
2. If someone wants to download something, provide `nightmare.download(selector|url)` which returns a buffer. if a buffer is not possible, support `path` in the signature to write to a file.
3. When we are downloading something, wait until that download is complete in order to resume.
4. Potentially support progress events via `nightmare.on('progress', fn)` if that's supported

This increases the API to:

```js
nightmare
  .on('progress', fn)
  .download('a[href="/segmentio/nightmare/archive/master.zip"]')
```

One thing we should also consider is... does nightmare even need to support downloads, or can we get away with:

```js
var request = require('request')
var fs = require('fs')

var url = yield nightmare
  .evaluate(function() {
    return document.querySelector('a[href="/segmentio/nightmare/archive/master.zip"]').getAttribute('href')
  })

request.get('url').pipe(fs.createWriteStream(__dirname + 'somepath.txt'))
```

Are there any advantages to having this in core?

**Comment by subcontact** [comment]
Firstly - awesome library and great pull request.
I wanted to share a scenario where I am logged in with a session and need to click on a button that runs some javascript that POSTS a form with a unique URL and data set and cannot only be requested once (each submission creates a new ID at the end of the URL so cannot be replayed).
The response is a download attachment (Content-Disposition: attachment).

I need to get get the response attachment somehow - an API would allow for that. Currently I haven't figured out how to do this.

Thanks

**Comment by matthewmueller** [comment]
@subcontact hmmm.. that's a tough one, have you tried getting it to work with the current PR?

**Comment by rosshinkley** [comment]
@matthewmueller Thank you for your reply, and I'm sorry I couldn't get back to this sooner.  Also, apologies in advance for the missive.  Responding to your feedback:

> By default ignore all downloads

Seems reasonable to me.  I can do that as time permits.

> If someone wants to download something, provide nightmare.download(selector|url) which returns a buffer. if a buffer is not possible, support path in the signature to write to a file.

I don't understand what you're asking, what should the selector provided do?  Click?  As for providing a URL, would that be a simple get?  And with the buffer, no, I don't think that is possible.  Electron supplies the filepath and a method to override the filepath.

> When we are downloading something, wait until that download is complete in order to resume.

Blocking on each download isn't how browsers work.  From a test automation standpoint, I can certainly understand the need, but from a purely automation standpoint, I'm not sure I agree.

> Potentially support progress events via nightmare.on('progress', fn) if that's supported

Yes, that's supported (and can be supported with the reduced API).

Re: getting away with a simple request: no, I don't think you want to do that.  If you're making an authenticated request for a resource (thinking of something like a request to a private link in say, Dropbox), you want to do that from within the context of Electron.  Copying the resource URI and issuing a plain GET will almost certainly result in failure.  This (I think, and I'm sorry if I'm misreading here) gets to what @subcontact's problem - authenticated downloads won't work with simple requests.

Re: keeping at least some form of downloads in core, that's a _great_ question.  My answer to that would be a qualified no.  How downloads are handled is probably going to be dependent on use case.  To that end, I think the blocking `.download()` and the non-blocking methods in this PR should be broken out into at least one (probably two?) plugins.  

The problem with that is binding to `will-download`.  A new method to hook into `will-download` with a custom handler could be exposed, or (with modification to pass a child handler) the custom event bindings in #367 could be used.  Thoughts?

I'm willing to reduce the API as requested, and if you think it's a good idea, I will close this PR and open a fresh one with those changes.

**Comment by matthewmueller** [comment]
@rosshinkley thanks for the detailed response!

> I don't understand what you're asking, what should the selector provided do? Click? As for providing a URL, would that be a simple get? And with the buffer, no, I don't think that is possible. Electron supplies the filepath and a method to override the filepath.

Yah, you're right, when I proposed that I was thinking about it in terms of what the click/mouseover/etc methods do, but that's not the same as download. 

> Blocking on each download isn't how browsers work. From a test automation standpoint, I can certainly understand the need, but from a purely automation standpoint, I'm not sure I agree.

I was proposing this to handle `wait('downloads-complete')` piece that you added in this PR. Haven't investigated how that would work internally though.

> Re: getting away with a simple request: no, I don't think you want to do that. If you're making an authenticated request for a resource (thinking of something like a request to a private link in say, Dropbox), you want to do that from within the context of Electron. Copying the resource URI and issuing a plain GET will almost certainly result in failure. This (I think, and I'm sorry if I'm misreading here) gets to what @subcontact's problem - authenticated downloads won't work with simple requests.

From my understanding, a POST request triggers the download, so whether this is a plugin or not, we'd need to at least know how to do that with core.

> Re: keeping at least some form of downloads in core, that's a great question. My answer to that would be a qualified no. How downloads are handled is probably going to be dependent on use case. To that end, I think the blocking .download() and the non-blocking methods in this PR should be broken out into at least one (probably two?) plugins.

Yah, I agree.

> The problem with that is binding to will-download. A new method to hook into will-download with a custom handler could be exposed, or (with modification to pass a child handler) the custom event bindings in #367 could be used. Thoughts?

Yah, I think we should add this regardless, we're already hooking into the other events. Seems like it would be as simple as:

`win.webContents.session.on('will-download', forward('will-download'))`

Then you can hook into it with an action:

```js
Nightmare.action('download', function () {
  this.child.on('will-download', handler)
})
```

Haven't thought through exactly how this would work, but that would be the approach I would investigate.

**Comment by rosshinkley** [comment]
@matthewmueller no problem. :)  Thanks for yours as well!

Tying up some loose ends:

> I was proposing this to handle `wait('downloads-complete')` piece that you added in this PR. Haven't investigated how that would work internally though.

Yeah, I understand the need, and also buy `.wait('downloads-complete')` is problematic.  Totally off the cuff: I think that enforcing one download per `.download()` call would probably work.  It would consist of (more or less) how it works now, but with an extra wait for `will-download` to be caught, and the internal method using the wait-download function.

> From my understanding, a POST request triggers the download, so whether this is a plugin or not, we'd need to at least know how to do that with core.

I was more concerned with the "authenticated" part, not with that it was triggered on POST.  I don't _think_ it should matter.  Electron doesn't seem to care what triggered a download, just that one happens.  That being said, it's probably worth propping up a minimal test case to be certain.

> Yah, I agree [to plugins].

The more I think about this, the more sense it makes, and maybe should redirect my attention to #391.  I was thinking about how to extend the Electron process earlier today, because, well...

> Seems like it would be as simple as: `win.webContents.session.on('will-download', forward('will-download'))`

Handling this appropriately seems to require doing some special steps in the Electron process.  In tinkering with `.on('will-download', fn)`, as soon as the `DownloadItem` fell out of the handler scope, Electron would execute whatever behavior had been defined by synchronous completion (almost like the `on` handler is more like middleware, if that makes sense).  The [Electron docs for `setSavePath`](https://github.com/atom/electron/blob/master/docs/api/download-item.md#downloaditemsetsavepathpath) seem to support what I found by experimentation:

>The API is only available in session's will-download callback function. If user doesn't set the save path via the API, Electron will use the original routine to determine the save path(Usually prompts a save dialog).

In other words, trying to do something via event communication would (almost) certainly result in a blocking download dialog.

I think I'm going to close this PR, backburner the download plugins for the moment, and direct my remaining questions to #391 as I think they're more sensible under that topic.

**Comment by AspireToCodeBetter** [comment]
@rosshinkley since the merge of this PR seems to be on hold. would you be able to provide some example to download a file using the internals? Sorry, I am new to both electron and nightmare

**Comment by rosshinkley** [comment]
@AspireToCodeBetter TL;DR: At present, you'd have to fork Nightmare and roll your own.

There is currently work being done to add extensions to the Electron process that Nightmare wraps (#391), one of the benefits of this being the ability to attach a handler to `will-download`.  The reason this PR was ultimately closed is because there are several use-cases for `will-download`, leading me to believe they should be broken out into plugins to be used with (as of this writing, incomplete) extending Electron.

When #391 is complete, I'll see to it that the download plugins are completed and put on NPM, and  certainly post a link on this thread and possibly on a few others (#151 and #371 come to mind).
